File: leetcode-js/297-serialize-and-deserialize-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    const result = [];
    buildString(root, result);
    return result.join(',');
};

function buildString(node, result) {
    if (!node) {
        result.push('X');
    } else {
        result.push(node.val);
        buildString(node.left, result);
        buildString(node.right, result);
    }
}

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    const queue = data.split(',');
    return buildTree(queue);
};

function buildTree(queue) {
    const val = queue.shift();
    if (val === 'X') {
        return null;
    } else {
        const node = new TreeNode(parseInt(val));
        node.left = buildTree(queue);
        node.right = buildTree(queue);
        return node;
    }
}

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

console.log(serialize(new TreeNode(1)))
// console.log(deserialize());

function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
}

File: leetcode-java/297-serialize-and-deserialize-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        preorder(root, sb);
        return sb.toString();
    }

    void preorder(TreeNode node, StringBuilder sb) {
        if (node == null) {
            sb.append('#').append(',');
        } else {
            sb.append(node.val).append(',');
            preorder(node.left, sb);
            preorder(node.right, sb);
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        Deque<String> queue = new LinkedList<>();
        queue.addAll(Arrays.asList(data.split(",")));
        return build(queue);
    }

    TreeNode build(Deque<String> queue) {
        String val = queue.remove();
        if (val.equals("#")) {
            return null;
        }
        TreeNode node = new TreeNode(Integer.valueOf(val));
        node.left = build(queue);
        node.right = build(queue);
        return node;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
